### PR TITLE
Table Preview Updates

### PR DESCRIPTION
**Pull request by themadcreator**
Separate mutable table into its own file.
Add session local storage for saving settings.
Add reset button to revert to defaults.

### CODE REVIEW

**Comment by llorca** [comment]
@themadcreator does this overlap with #1603? Also instead of a reset button, what do you think about https://github.com/palantir/blueprint/pull/1603#issuecomment-331560009

File: packages/table/preview/localStore.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
+/**
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ * Licensed under the BSD-3 License as modified (the “License”); you may obtain a copy
+ * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+ * and https://github.com/palantir/blueprint/blob/master/PATENTS
+ *
+ * Demonstrates sample usage of the table component.
+ */
+
+import { defaults } from "lodash";
```
Comment (cmslewis) [inline packages/table/preview/localStore.ts]: I see `lodash` is only a dev dependency, but still, do we really need it for this one function? Would prefer not to pull it in if we've gone this long without it.

Alternatively, we could just cherry-pick the function we need, as mentioned in their docs:

```tsx
// Cherry-pick methods for smaller browserify/rollup/webpack bundles.
var at = require('lodash/at');
var curryN = require('lodash/fp/curryN');
```

File: packages/table/preview/localStore.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+export class LocalStore<T> {
+    private storage: Storage;
+
+    constructor(private key: string, session = false){
+        this.storage = session ? sessionStorage : localStorage;
+    }
+
+    public getWithDefaults(defs?: T): T | {} {
```
Comment (cmslewis) [inline packages/table/preview/localStore.ts]: Would prefer to call this `defaults`. `defs` means "definitions" to me.

File: packages/table/preview/mutableTable.tsx [hunk 3]
```diff
@@ -0,0 +1,969 @@
...
+function getRandomInteger(min: number, max: number) {
+    // min and max are inclusive
+    return Math.floor(min + Math.random() * (max - min + 1));
+}
+
+function contains(arr: any[], value: any) {
+    return arr.indexOf(value) >= 0;
+}
+
+const DEFAULT_STATE: IMutableTableState = {
```
Comment (cmslewis) [inline packages/table/preview/mutableTable.tsx]: Can we define this _below_ the `IMutableTableState` interface? Keeps the `DEFAULTS` definition closer to its usage in the constructor, and that order would just make more sense to me.

File: packages/table/preview/localStore.ts (hunk 2 above)
Comment (themadcreator) [reply packages/table/preview/localStore.ts]: name already taken by the lodash method, but i'll come up w somethign

File: packages/table/preview/localStore.ts (hunk 1 above)
Comment (themadcreator) [reply packages/table/preview/localStore.ts]: not important for preview, but i'll switch it

File: packages/table/preview/localStore.ts (hunk 2 above)
Comment (cmslewis) [reply packages/table/preview/localStore.ts]: Ah, boo.

File: packages/table/preview/localStore.ts (hunk 1 above)
Comment (themadcreator) [reply packages/table/preview/localStore.ts]: when I switched to `import * as defaults from "lodash/defaults"` the types couldn't be resolved.

**Pull request merged**